### DISCUSSION TITLE
[mini.surround] how to do conditional surrounding based on input?

### DISCUSSION THREAD

**Discussion by aca** [nvim-mini/mini.nvim#462] [Q&A] [answered]
Hi, I'm trying to migrate my config from vim-sandwich to mini. I hope I could get some help.

I want to make 'c' to add code block in markdown.

- If the target to surround is single line, surround with "`". 
- If the target is multi line surround with "```". 

```
asd -> `asd`

asd\nasd -> ```asd\nasd```
```

I'm not sure how to do "GETINPUT()" 
```
["c"] = {
    output = function(a)
        local content = GETINPUT()
        if string.find(content, "\n") == nil then
            return { left = "`", right = "`" }
        else
            return { left = "```\n", right = "\n```\n" }
        end
    end,
},
```

**Comment by echasnovski** [comment, marked as answer]
There is no easy method to do this, because callable `output` specification gets called without arguments and no guarantees at what Neovim state it does so. But it is an interesting use case.

Here is my closest attempt:
```lua
require('mini.surround').setup({
  custom_surroundings = {
    c = {
      output = function()
        local is_visual = _G.is_visual
        _G.is_visual = nil
        local mark_from = is_visual and "'<" or "'["
        local mark_to = is_visual and "'>" or "']"
        local n_content_lines = vim.fn.line(mark_to) - vim.fn.line(mark_from)
        if n_content_lines == 0 then return { left = '`', right = '`' } end
        return { left = '```\n', right = '\n```\n' }
      end,
    },
  },
})

local rhs = [[<Cmd>lua _G.is_visual=true<CR>:<C-u>lua MiniSurround.add('visual')<CR>]]
vim.keymap.set('x', 'sa', rhs, { silent = true })
```

Quick notes:
- It doesn't compute whole content per se, but directly decides whether it is single or multiple line. It does so by utilizing information from marks depending on mode.
- As it seems to always be executed in Normal mode, there needs to be a way to distinguish when adding is performed in Visual or Operator-pending mode. This is the cause of all the `_G.is_visual` trickery. If you decide to only apply this in Visual mode, then no need for that.
- It doesn't really work when multiline text is indented. It is by design of 'mini.surround' that it will place left surrounding part at the left of the first non-blank character in linewise mode. You can try `config.respect_selection_type = true` and `{ left = '```', right = '```' }` as returned value, but it will indent the code (again, by design to account for more common use cases).


------

Another solution would be to make a prompt for user to add extra key press and tell which type of code surrounding they want. For example, with `vim.fn.getcharstr()` and information message with `print()`.

------

That said, my **proper** solution would be to advise you to use separate surroundings for single and multi line cases. The first one is plain `` ` ``, the second one can be mapped to `c`.

**Reply by aca** [reply]
Thanks!

By the way, respect_selection_type = true, adds wierd character `^@`,  is it bug?

    ```�
        wer
        wer
    �```�

**Reply by echasnovski** [reply]
> Thanks!
> 
> By the way, respect_selection_type = true, adds wierd character `^@`, is it bug?
> 
> ```
> ```�
>     wer
>     wer
> �```�
> ```

Those are `\n` characters and not needed. Use `{ left = '```', right = '```' }` directly.

**Comment by aca** [comment]
Thanks! 

I ended up with this to make it work with indent.
This is so messy, but just for the record if anyone needs it.

```
        ["c"] = {
            -- https://github.com/echasnovski/mini.nvim/discussions/462
            output = function()
                -- is_visual doesn't work
                -- local is_visual = vim.tbl_contains({ 'v', 'V', '\22' }, vim.fn.mode())
                -- local mark_from = is_visual and "'<" or "'["
                -- local mark_to = is_visual and "'>" or "']"
                local mark_from = "'<"
                local mark_to = "'>"
                local startline = vim.fn.line(mark_from)
                local endline = vim.fn.line(mark_to)
                local n_content_lines = endline - startline
                if n_content_lines == 0 then
                    return { left = "`", right = "`" }
                end
                local firstline = vim.api.nvim_buf_get_lines(0, startline - 1, startline, false)
                local _, indent = string.find(firstline[1], '^%s*')
                if indent ~= nil then
                    return { left = "```\n" .. string.rep(" ", indent) , right = "\n" .. string.rep(" ", indent) .. "```\n" }
                else
                    return { left = "```\n", right = "\n```\n" }
                end
            end,
        },
```
